Zend/Optimizer/zend_inference.c

```diff
@@ -3024,7 +3024,14 @@ static zend_always_inline zend_result _zend_update_type_info(
 			break;
 		case ZEND_ASSIGN_OBJ:
 			if (opline->op1_type == IS_CV) {
-				tmp = (t1 & (MAY_BE_REF|MAY_BE_OBJECT))|MAY_BE_RC1|MAY_BE_RCN;
+				zend_class_entry *ce = ssa_var_info[ssa_op->op1_use].ce;
+				bool add_rc = !ce
+					|| ce->__set
+					/* Non-default write_property may be set within create_object. */
+					|| ce->create_object
+					|| ce->default_object_handlers->write_property != zend_std_write_property
```

Comment:
I think this line is not necessary. In case ``create_object`` is NULL, PHP uses ``std_object_handlers`` and therefore ``zend_std_write_property``.

Reply:
https://github.com/php/php-src/blob/34e2dc5697283dfa01c5f85cceafb99bb372ac7c/Zend/zend_objects.c#L34

`default_object_handlers` is initialized during startup, so it seems to me `create_object` is not necessary for setting `write_property` to something other than `zend_std_write_property`?

Reply:
I forgot ``ce->default_object_handlers`` was introduced in 8.3.
Actually, that made all the exiting checks for ``ce->create_object`` incomplete.
All of them need to be revisited and fixed

Now it seems, you really need to check only ``ce->default_object_handlers->write_property != zend_std_write_property``

cc @bwoebi

Reply:
@dstogov I think what you said previously is still true. `create_object` could manage a copy of the handlers list and modify it at runtime, even just for specific objects. If that's not something we want to support, I think this should be documented explicitly and possibly verified in some place in debug builds.

Reply:
Now I'm not sure that class without ``create_object`` will necessary use ``std_object_handlers``.

Reply:
@dstogov I mean, in theory `zend_object.handlers` can be replaced anywhere. Realistically, it should only be replaced in `create_object`. `ce->create_object || ce->default_object_handlers->write_property != zend_std_write_property` seems like a safe assumption to me, as then either it would have been modified through `default_object_handlers`, or in `create_object`.

Reply:
You check is right. I think the check for ``create_object`` is even redundant now.

I talk about Optimizer code consistency. In different places Optimizer checks only for ``create_object`` and this checks became incorrect after ``ce->default_handlers`` introduction.
It would be better to fix that code first.

Reply:
I understand, I'll have a look at the existing checks. What I was referencing was whether we still consider it valid to switch out `zend_object.handlers` on-the-fly in `create_object`, diverging from `ce->default_object_handlers`.